File: src/services/weatherApi.js
const API_KEY = import.meta.env.VITE_OPENWEATHER_API_KEY;
const BASE_URL = 'https://api.openweathermap.org/data/2.5';
const GEO_URL = 'https://api.openweathermap.org/geo/1.0';

export const fetchWeatherByCoordinates = async (lat, lon) => {
  if (!API_KEY) throw new Error('API Key is missing');
  const res = await fetch(`${BASE_URL}/weather?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric`);
  if (!res.ok) throw new Error('Failed to fetch weather data');
  return res.json();
};

export const fetchForecastByCoordinates = async (lat, lon) => {
  if (!API_KEY) throw new Error('API Key is missing');
  const res = await fetch(`${BASE_URL}/forecast?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric`);
  if (!res.ok) throw new Error('Failed to fetch forecast data');
  return res.json();
};

export const fetchCoordinatesByCity = async (cityName) => {
  if (!API_KEY) throw new Error('API Key is missing');
  const res = await fetch(`${GEO_URL}/direct?q=${cityName}&limit=1&appid=${API_KEY}`);
  if (!res.ok) throw new Error('Failed to fetch coordinates');
  const data = await res.json();
  if (data.length === 0) throw new Error('City not found');
  return data[0]; // Returns { name, lat, lon, country, state }
};

export const fetchWeatherByCityId = async (id) => {
  if (!API_KEY) throw new Error('API Key is missing');
  const res = await fetch(`${BASE_URL}/weather?id=${id}&appid=${API_KEY}&units=metric`);
  if (!res.ok) throw new Error('Failed to fetch weather data');
  return res.json();
};

export const fetchHistoricalWeather = async (lat, lon) => {
  // Use Open-Meteo for free historical weather data (last 7 days)
  const today = new Date();
  const endDate = today.toISOString().split('T')[0];
  
  const startDateObj = new Date(today);
  startDateObj.setDate(today.getDate() - 7);
  const startDate = startDateObj.toISOString().split('T')[0];

  const res = await fetch(
    `https://archive-api.open-meteo.com/v1/archive?latitude=${lat}&longitude=${lon}&start_date=${startDate}&end_date=${endDate}&hourly=temperature_2m`
  );
  if (!res.ok) throw new Error('Failed to fetch historical weather');
  return res.json();
};

File: weather_app/lib/services/weather_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../env.dart';
import '../models/weather_model.dart';

class WeatherApi {
  static const String _baseUrl = 'https://api.openweathermap.org/data/2.5';
  static const String _geoUrl = 'https://api.openweathermap.org/geo/1.0';

  static String _parseError(http.Response res, String fallback) {
    try {
      final json = jsonDecode(res.body);
      return json['message']?.toString() ?? fallback;
    } catch (_) {
      return fallback;
    }
  }

  static Future<WeatherData> fetchWeatherByCoordinates(double lat, double lon) async {
    if (Env.openWeatherApiKey.isEmpty || Env.openWeatherApiKey == 'YOUR_API_KEY_HERE') {
      throw Exception('OpenWeatherMap API Key is missing. Please add it to env.dart');
    }
    final res = await http.get(Uri.parse('$_baseUrl/weather?lat=$lat&lon=$lon&appid=${Env.openWeatherApiKey}&units=metric'));
    if (res.statusCode != 200) throw Exception(_parseError(res, 'Failed to fetch weather data'));
    return WeatherData.fromJson(jsonDecode(res.body));
  }

  static Future<List<ForecastData>> fetchForecastByCoordinates(double lat, double lon) async {
    if (Env.openWeatherApiKey.isEmpty || Env.openWeatherApiKey == 'YOUR_API_KEY_HERE') {
      throw Exception('OpenWeatherMap API Key is missing. Please add it to env.dart');
    }
    final res = await http.get(Uri.parse('$_baseUrl/forecast?lat=$lat&lon=$lon&appid=${Env.openWeatherApiKey}&units=metric'));
    if (res.statusCode != 200) throw Exception(_parseError(res, 'Failed to fetch forecast data'));
    
    final data = jsonDecode(res.body);
    final List list = data['list'];
    
    // Filter to one per day (~noon)
    return list
        .where((item) => (item['dt_txt'] as String).contains('12:00:00'))
        .map((item) => ForecastData.fromJson(item))
        .take(5)
        .toList();
  }

  static Future<GeoData> fetchCoordinatesByCity(String cityName) async {
    if (Env.openWeatherApiKey.isEmpty || Env.openWeatherApiKey == 'YOUR_API_KEY_HERE') {
      throw Exception('OpenWeatherMap API Key is missing. Please add it to env.dart');
    }
    final res = await http.get(Uri.parse('$_geoUrl/direct?q=$cityName&limit=1&appid=${Env.openWeatherApiKey}'));
    if (res.statusCode != 200) throw Exception(_parseError(res, 'Failed to fetch coordinates'));
    
    final List data = jsonDecode(res.body);
    if (data.isEmpty) throw Exception('City not found');
    return GeoData.fromJson(data[0]);
  }

  static Future<HistoryData> fetchHistoricalWeather(double lat, double lon) async {
    final today = DateTime.now();
    final endDate = today.toIso8601String().split('T')[0];
    
    final startDateObj = today.subtract(const Duration(days: 7));
    final startDate = startDateObj.toIso8601String().split('T')[0];

    final res = await http.get(Uri.parse(
        'https://archive-api.open-meteo.com/v1/archive?latitude=$lat&longitude=$lon&start_date=$startDate&end_date=$endDate&hourly=temperature_2m'));
    
    if (res.statusCode != 200) throw Exception('Failed to fetch historical weather');
    return HistoryData.fromJson(jsonDecode(res.body));
  }
}
